plugins/mewbot-discord_dice_roller/src/mewbot_discord_dice_roller/triggers.py

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python3
+# pylint: disable=duplicate-code
+# this is an example - duplication for emphasis is desirable
+
+"""
+Triggers every time a discord message is created.
+"""
+
+from __future__ import annotations
+
+from typing import Set, Type
+
+from mewbot.api.v1 import Trigger
+from mewbot.core import InputEvent
+from mewbot.io.discord import DiscordMessageCreationEvent
+
+
+class DiscordDiceRollCommandTrigger(Trigger):
+    """
+    Fires every time a DiscordMessageCreationEvent has text that starts with "!roll".
+    """
+
+    @staticmethod
+    def consumes_inputs() -> Set[Type[InputEvent]]:
+        """
+        Triggers off all DiscordMessageCreationEvents.
+
+        :return:
+        """
+        return {DiscordMessageCreationEvent}
+
+    def matches(self, event: InputEvent) -> bool:
+        """
+        Matches all DiscordMessageCreationEvent with the starting keyword !roll.
+
+        :param event:
+        :return:
+        """
+
+        if not isinstance(event, DiscordMessageCreationEvent):
```

Comment:
Idle thought for future code; is there a way to make sure this happens automatically/is not forgettable by a user?
Is there a reason not to do that?

Reply:
The [Liskov substitution principle](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides) boils down to the argument to the function being `InputEvent` (or `OutputEvent` for `Output`s), so the check is going to be needed for typing purposes. It's also just a good defensive measure against poorly implemented bot runners.

It _is_ a little cumbersome, and we could abstract it to a decorator...
A fully generic version might look like

```python
from typing import Callable, ParamSpec, TypeVar

import mewbot.api.v1
from mewbot.core import InputEvent, OutputEvent
from mewbot.io.common import EventWithReplyMixIn


Args = ParamSpec("Args")
Bound = TypeVar("Bound", bound=InputEvent)
Return = TypeVar("Return")


def check_types(*types: type[Bound], result: Return = ...) -> Callable[[Callable[Args, Return]], Callable[Args, Return]]:
    def check_wrapper(wrapped: Callable[Args, Return]) -> Callable[Args, Return]:
        def inner(*args: Args.args, **kwargs: Args.kwargs) -> Return:
            print(args)
            if not isinstance(args[1], types):
                if result is ...:
                    raise TypeError(f"First argument to {wrapped} must be one of {types}")
                return result

            return wrapped(*args, **kwargs)

        return inner
    return check_wrapper


class A(mewbot.api.v1.Trigger):
    @staticmethod
    def consumes_inputs() -> set[type[InputEvent]]:
        return {EventWithReplyMixIn}

    @check_types(EventWithReplyMixIn, result=False)
    def matches(self, event: InputEvent) -> bool:
        return True
```

I suspect one which is configured specifically for the different functions would, however, be more useful (and not make mypy scream)

... ok, I took some time to play and it is 100% possible:

```python
from __future__ import annotations

from typing import Callable, TypeVar

from mewbot.api.v1 import Trigger, Condition, InputEvent
from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent

Component = TypeVar("Component", bound=Trigger | Condition)
Event = TypeVar("Event", bound=InputEvent)


def check_types(wrapped: Callable[[Component, Event], bool]) -> Callable[[Component, InputEvent], bool]:
    def inner(self: Component, event: InputEvent) -> bool:
        if not isinstance(event, tuple(self.consumes_inputs())):
            return False

        return wrapped(self, event)  # type: ignore

    return inner


class A(Trigger):
    @staticmethod
    def consumes_inputs() -> set[type[InputEvent]]:
        return {DiscordInputEvent}

    @check_types
    def matches(self, event: DiscordInputEvent) -> bool:
        print(event)
        return True


print("invoking A()")
print(A().matches(InputEvent()))
print(A().matches(DiscordInputEvent()))
print(A().matches(DiscordUserJoinInputEvent("nya")))
```

That passes typing checks; there is one ignore there because I can't _prove_ that `Bound` is implied by the `isintance` call above.

Reply:
...if we want to get really, really clever, we could actually shift the concern of `consume_inputs` to the type signature of `matches`...

I think this is where it veers into bad idea territory. And I can't easily make it work, but I have made a slightly better version of the above (which doesn't need any mypy suppressions, but *does* confuse PyCharm a bit):
```python
from __future__ import annotations

import functools

from types import UnionType
from typing import Callable, TypeVar, get_type_hints

from mewbot.api.v1 import Trigger, Condition, InputEvent

Component = TypeVar("Component", bound=Trigger | Condition)
Event = TypeVar("Event", bound=InputEvent)


def check_types(
    wrapped: Callable[[Component, Event], bool]
) -> Callable[[Component, InputEvent], bool]:
    types = get_type_hints(wrapped)
    if "event" not in types:
        raise TypeError("Received function without 'event' parameter")

    event_types: type[Event] = types["event"]
    if not isinstance(event_types, (type, UnionType)):
        raise TypeError("Received function with invalid type for 'event'")

    @functools.wraps(wrapped)
    def inner(self: Component, event: InputEvent) -> bool:
        # noinspection PyTypeHints
        if not isinstance(event, event_types):
            return False

        return wrapped(self, event)

    return inner


# Demo Code

from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent
from mewbot.io.http import IncomingWebhookEvent


class A(Trigger):
    @staticmethod
    def consumes_inputs() -> set[type[InputEvent]]:
        return {DiscordInputEvent, IncomingWebhookEvent}

    @check_types
    def matches(self, event: DiscordInputEvent | IncomingWebhookEvent) -> bool:
        """Hello I am documentation"""
        return True


print(A.matches)  # <function A.matches at 0x7f6e703f4a40>
print(A.matches.__doc__)  # <function A.matches at 0x7f6e703f4a40>
print(A().matches(InputEvent()))  # False
print(A().matches(DiscordInputEvent()))  # True
print(A().matches(DiscordUserJoinInputEvent("nya")))  # True
```

And I think we could store the `UnionType` somewhere and use it for `consume_inputs` if we really wanted to. But I also feel that might be Too Much Magic:tm:

Reply:
I really like this. It's very elegant and could be more magical.

This feels like it should be it's own pull request. I'll do one tomorrow to get it into the code base and get it integrated.